### ISSUE TITLE
Plugin fails to get metadata if java/scala compilation fails

### ISSUE THREAD

**Issue by or-shachar** [bazelbuild/intellij#1167]
Using bazel scala project we noticed that if compilation fails the plugin fails to get the metadata.

When we try to sync currently the plugin runs 2 bazel commands:
```
- bazel info ...
- bazel build ... --output_groups=intellij-info-generic,intellij-info-java,intellij-resolve-java
```

We noticed that if we remove the `intellij-resolve-java` the second operation will succeed.

We should probably break the second bazel invocation to two phases so we will still get the metadata even if the compilation fails.

cc: @jin @chaoren @ittaiz

**Comment by brendandouglas** [comment]
There are a few reasons for the current situation:

- running bazel build invocations has a very large overhead internally (7-8 seconds when run remotely, before any bazel work is actually done), so we want to minimize the number of builds.

- there's no clear distinction between the 'metadata' (target graph, target details) and the output of the resolve step (compilation output, in the case of java and scala). They're co-dependent, so updating the metadata without resolve output can lead to bad behavior.

Instead, we assume that the 'resolve' build will succeed initially (for instance, because the user is checking out code without local modifications). Any subsequent builds for which some targets fail to compile will result in that target's 'metadata' and compilation output remaining unchanged, which is generally the desired outcome. We'd need a very good reason to consider changing this approach.

**Comment by ittaiz** [comment]
Thanks for responding!
Can you elaborate when/how `updating the metadata without resolve output can lead to bad behavior`? Not sure I understand why that is and for a very very long time the two commands were indeed separate.

I think part of the problem in the above assumption is that it's a bit too clean. Sometimes someone will rebase onto master they'll have a non compiling file. If this file is in the "middle" of the graph then all dependent targets will be messed up even if they are new from master. This is something that's not trivial to track down without deeply investigating the tool or really understanding Bazel.
We'd love to give users a simpler experience then what is now expected from people.

**Comment by brendandouglas** [comment]
I did combine the two build steps recently (a few months ago) for performance reasons, but I can't think of a situation where that would actually cause a change in behavior like you're seeing.

We request all the build outputs in the same build invocation, but we still *handle* the 'info' and 'resolve' outputs separately -- that hasn't changed. If individual targets fail to build, we should still be updating their metadata in the project data.

It might help to have a small reproducer here (ideally in java if possible).

**Comment by ittaiz** [comment]
I'll add a repro no problem. Main issue is where `A->B->C` and `B` doesn't compile. Then you don't get `C`'s `info` outputs (of course it could be a very long and complicated sub tree of targets)

**Comment by ittaiz** [comment]
repro repo: https://github.com/ittaiz/bazel-ij-info-with-resolve 
After sync `B` and `A` are unsynced:
<img width="403" alt="Screen Shot 2019-09-20 at 18 04 40" src="https://user-images.githubusercontent.com/1378327/65337860-a8e8fd00-dbd1-11e9-8ab5-715bc41b966d.png">
Because of a compilation issue in `B`:
<img width="424" alt="Screen Shot 2019-09-20 at 18 04 54" src="https://user-images.githubusercontent.com/1378327/65337859-a8e8fd00-dbd1-11e9-9adf-583f5060f5bb.png">
As you can see the external dependencies of `B` and of `A` aren't resolved:
<img width="397" alt="Screen Shot 2019-09-20 at 18 05 10" src="https://user-images.githubusercontent.com/1378327/65337856-a8e8fd00-dbd1-11e9-8d8a-6abc0b682920.png">

**Comment by jin** [comment]
The current behavior seems to fit my intuition again. It all boils down to this:

`@guava` is a dependency of B, and only B in the entire project. If `B` fails to provide the `intellij-resolve-java` output group requested by the aspect, the aspect will not continue traversing the dep edge up to `@guava`. This means that the target map will not contain any information about `@guava`.

However, if you add `@guava` as a dep of C, and because C compiles, the aspect can travel to `@guava` and complete the `intellij-resolve-java` output group. Now, even if `B` doesn't compile, its `import com.google.common.cache.AbstractCache` import statement is no longer red. 

This has nothing to do with the difference between `intellij-info-java` and `intellij-resolve-java`. `intellij-info-java` doesn't provide the information required to fix the red code issue. 

Interestingly, even though `B` is the only rdep of `@guava` and `B`'s compilation failed, the ijar and intellij-info.txt are created:

```
$ tree bazel-bin/external/com_google_guava_guava/
bazel-bin/external/com_google_guava_guava/
├── com_google_guava_guava-46378293.intellij-info.txt
└── _ijar
    └── com_google_guava_guava
        └── external
            └── com_google_guava_guava
                └── guava-20.0-ijar.jar
```

So the fix here should be adding `@guava` into the target index even if all its callers did not compile and sync successfully.

**Comment by brendandouglas** [comment]
Thanks for the repro project, I see the problem.

The part I was missing is that the intellij-info.txt output artifact is produced (expected), but isn't declared in the BEP output (unexpected).

This feels like a bug in BEP. Prior to BEP we relied on parsing stdout/stderr, with the --experimental_show_artifacts flag set. The intellij-info outputs for B are declared just fine there, regardless of whether the jars are successfully compiled.

**Comment by jin** [comment]
@brendandouglas are you saying the root cause is that these two lists aren't consistent with each other?

```
$ tree -fi bazel-bin/ | grep intellij-info.txt
bazel-bin/A-65.intellij-info.txt
bazel-bin/B-66.intellij-info.txt
bazel-bin/C-67.intellij-info.txt
bazel-bin/external/bazel_tools/tools/jdk/current_java_toolchain--1286701614.intellij-info.txt
bazel-bin/external/com_google_code_findbugs_jsr305/com_google_code_findbugs_jsr305-1219275564.intellij-info.txt
bazel-bin/external/com_google_errorprone_error_prone_annotations/com_google_errorprone_error_prone_annotations--1327292041.intellij-info.txt
bazel-bin/external/com_google_guava_guava/com_google_guava_guava-46378293.intellij-info.txt
bazel-bin/external/google_java_format/google_java_format--2076037202.intellij-info.txt
bazel-bin/external/remote_java_tools_linux/toolchain--412636119.intellij-info.txt
```

```
$ cat /tmp/f.txt | grep intellij-info.txt
    name: "external/remote_java_tools_linux/toolchain--412636119.intellij-info.txt"
    uri: "file:///usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/af0112346abd86e45930298bb524e045/execroot/main/bazel-out/k8-fastbuild/bin/external/remote_java_tools_linux/toolchain--412636119.intellij-info.txt"
    name: "external/bazel_tools/tools/jdk/current_java_toolchain--1286701614.intellij-info.txt"
    uri: "file:///usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/af0112346abd86e45930298bb524e045/execroot/main/bazel-out/k8-fastbuild/bin/external/bazel_tools/tools/jdk/current_java_toolchain--1286701614.intellij-info.txt"
    name: "C-67.intellij-info.txt"
    uri: "file:///usr/local/google/home/jingwen/.cache/bazel/_bazel_jingwen/af0112346abd86e45930298bb524e045/execroot/main/bazel-out/k8-fastbuild/bin/C-67.intellij-info.txt"
```

**Comment by brendandouglas** [comment]
Yes, if that second list corresponds to the BEP contents

**Comment by jin** [comment]
Thanks, filed https://github.com/bazelbuild/bazel/issues/9413

**Comment by ittaiz** [comment]
Interesting, thanks!
Just to close the Bazel understanding gap- I expected A's intellij-info files not to be created because B failed. Are they created because these are in reality two different graphs (one for the resolve-java actions and one for the intellij/java-info actions)?

**Comment by ittaiz** [comment]
Also,
@brendandouglas thanks for iterating with us on this. It indeed seems like the BEP fix is better for everyone (blaze and bazel). Is there a cheap workaround you think we can do in the meanwhile (I know you're not a fan of temporary issues)?

**Comment by brendandouglas** [comment]
Thanks for filing the Bazel-side bug @jin.

@ittaiz the intellij-info outputs are not dependent on the compilation outputs, so in theory they could be implemented with different graphs. In practice I doubt they are though.

There's no cheap workaround, so I'll wait to hear back from the Bazel team. If they're willing to support this use case, I'll leave as is and wait for a Bazel-side fix. If not, I'll conditionally separate the 'info' and 'resolve' steps.

**Comment by jin** [comment]
> the intellij-info outputs are not dependent on the compilation outputs, so in theory they could be implemented with different graphs. In practice I doubt they are though.

Technically they are different subgraphs in the Skyframe-sense, since the intellij-info outputs were requested by a different output group from the compilation output group. 

I'll look into this issue. It's a particularly specific but important edge case for IDE use cases.

**Comment by jin** [comment]
Digging into this a little deeper, this issue only seems to happen if you're doing an **initial sync** and that the Java code doesn't compile.

If you've successfully performed an initial sync, and then you sync again even if a compilation (e.g. B in the repro project), the metadata of the rest of the files don't get affected. 

@ittaiz @or-shachar could you please confirm how often the aforementioned situation happens?

**Comment by ittaiz** [comment]
@jin I'm not sure I understand. Is the scenario you're describing about whether or not this issue also discards existing metadata of previously resolved and working files (`A`)? 
I think you're right. However if there are any changes (for example pulled from upstream) then those aren't synced and are broken. Is the use-case clear?

**Comment by ittaiz** [comment]
@brendandouglas @jin if Bazel won't support this use-case what do you think about running the `info` command only if the current command fails? This way most times people will still get the perf benefit. Additionally the change if they will support the use-case later on will just be dropping this extra call.

**Comment by jin** [comment]
The info command always runs before the command that builds the aspect. Do you mean separating out `intellij-info-java` into a separate command if `intellij-resolve-java` fails?

**Comment by ittaiz** [comment]
Yes, exactly

**Comment by ittaiz** [comment]
@brendandouglas wdyt about adding a fallback `bazel build` invocation without `intellij-resolve-java` if the first fails? Sounds simple and without a perf hit in most cases.
Main motivation is that we have a feature where you can add classes from the monorepo which IntelliJ doesn't currently know (using IntelliJ's smart auto complete). This feature is severely broken since we currently:

1. Add the symbol chosen by the user to the editor and also to the target.
2. Tell the user to comment out the symbol and IJ.
3. Trigger partial sync on the file.
4. Comment back in.
5. Use alt+enter to add an import.

What we'd like is:
1. Add the symbol chosen by the user to the editor and also to the target.
2. Programmatically trigger partial sync on the file. 
3. Use alt+enter to add an import.

For the user it's to go from 5 operations to 2 where often people have a hard time because they forget to comment out the symbol and so the sync fails and they waste some time on this.  

We'd be really interested in contributing this change if it's ok with you.

**Comment by brendandouglas** [comment]
I'd prefer to push for a Bazel-side fix here, rather than adding more complexity to the sync process. I've pinged the bazel-side bug and filed an internal bug for this issue -- let's see what the response is.

In the meantime, there's one part of that situation I don't understand. If both the source file and target deps are updated in step 1, why does the build fail in step 3 (without step 2)?

**Comment by ittaiz** [comment]
Good question (and sorry for missing the notification)- what we add in step 1 is merely the symbol in the editor.
For example imagine someone adding a new variable `Foo a` and while they're typing `Fo` nothing is recognized so they use smart autocomplete and what we do is complete `Fo` to `Foo`.
We don't add the import above because we need a PsiElement to do that correctly without messing with imports and such and so compilation fails (`Foo` doesn't yet have an import).
Does that make sense?

**Comment by brendandouglas** [comment]
Thanks, that workflow makes sense. Though I don't understand how the ide-info data helps.

I'd expect that the target is already synced prior to step 1, so adding 'Foo' to the source file then syncing wouldn't change anything (you'd still be missing both the import and dep and syncing wouldn't add them)

**Comment by ittaiz** [comment]
The difference would have been in IJ-Bazel plugin "discovering" about `@foo` and configuring it as an IntelliJ library. This in turn means that IntelliJ will index that jar and suggest the user to click alt+enter to add the import.
The idea is indeed that everything is synced prior to step 1 and then in it we add `Foo` to the editor and `@foo` to the `deps` of the relevant target.

If you think I'm missing something I'd love to know (I'm fairly certain this used to work for us before the unification to one invocation)

**Comment by ittaiz** [comment]
Also a separate use-case (and the one that caused us to look at this issue) is when you're working on prod code A and it doesn't compile for some reason  and you're switching to the test target ATest to add some code that requires a new dependency you need to "blindly" code in the editor because "new-test-dep" won't be identified until you stash changes to A and probably to ATest as well.
I'm re-surfacing this because one might suggest to just add the import manually somehow in the autocomplete scenario above but that will only solve one part of our needs.
Thanks in advance!

**Comment by brendandouglas** [comment]
> The difference would have been in IJ-Bazel plugin "discovering" about @foo and configuring it as an IntelliJ library

This is the bit I'm missing. Why would adding an invalid reference to 'Foo' then building cause the plugin to discover that @foo is missing? (it would be great if it did, I just haven't seen that happen, and can't see a mechanism for it)

**Comment by ittaiz** [comment]
Let's take the following case-
```
com
  wix
      biz
         BUILD.bazel (scala_library with name "biz" with glob *.scala and no deps)
         A.scala
```
and `A.scala` has:
```
package com.wix.biz
class A {
}
```
If I now go to `//com/wix/biz` target and add to it's deps `@foo` (assuming this is some kind of jvm_import_external or whatever) and then I run sync.
AFAIU Bazel IntelliJ plugin then `intellij-info-java` will find `@foo` as a compile time dependency of `//com/wix/biz` and so define it as an external library, right?
I say this because this is basically what we tell people to do (with the comment out part).
The problem the bazel bug + one invocation brings us is that the plugin never finds out about `@foo` is there is a broken target compilation wise (`intellij-resolve-java`)

If something still doesn't make sense please don't hesitate to ask.

**Comment by brendandouglas** [comment]
I see, I missed the part where the user manually updates the target deps. In that case, as a simpler workaround couldn't you change from:

1. Add the symbol to the editor and also to the target.
2. Programmatically trigger partial sync on the file.
3. Use alt+enter to add an import.

To:

1. Add the dep to the target only.
2. Programmatically trigger partial sync on the file.
3. Add the symbol to editor then Use alt+enter to add an import.

**Comment by brendandouglas** [comment]
In any case, hopefully this will be fixed on the bazel side. I believe they're currently working on it.

**Comment by ittaiz** [comment]
I think there's still a misunderstanding here- I'll clarify but if you'd rather we stop discussing it that's completely fine, please say so :)

The user doesn't manually update the target deps. Our plugin does.
The flow is that the **user** uses [smart autocomplete](https://www.jetbrains.com/help/idea/auto-completing-code.html#smart_completion) which adds an unresolved symbol to the editor. Our **plugin** programatically and implicitly adds the appropriate dependency and needs to trigger partial sync.

**Comment by brendandouglas** [comment]
Thanks, I understand now :)

**Comment by uri-canva** [comment]
Fix is in bazel 2.2.0: https://github.com/bazelbuild/bazel/commit/ecd2a5766233c088b70b69f3ce1b7a78834ceef9
